core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -409,20 +410,31 @@ TranslogWriter createWriter(long fileGeneration) throws IOException {
     public Location add(final Operation operation) throws IOException {
         final ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(bigArrays);
         try {
-            final BufferedChecksumStreamOutput checksumStreamOutput = new BufferedChecksumStreamOutput(out);
             final long start = out.position();
             out.skip(Integer.BYTES);
-            writeOperationNoSize(checksumStreamOutput, operation);
+            writeOperationNoSize(new BufferedChecksumStreamOutput(out), operation);
             final long end = out.position();
             final int operationSize = (int) (end - Integer.BYTES - start);
             out.seek(start);
             out.writeInt(operationSize);
             out.seek(end);
             final ReleasablePagedBytesReference bytes = out.bytes();
+            final Location location;
             try (ReleasableLock ignored = readLock.acquire()) {
                 ensureOpen();
-                return current.add(bytes, operation.seqNo());
+                location = current.add(bytes, operation.seqNo());
+            }
+            if (shouldFoldGeneration() && foldingGeneration.compareAndSet(false, true)) {
+                // we have to check the condition again lest we could fold twice in a race
+                if (shouldFoldGeneration()) {
+                    try (ReleasableLock ignored = writeLock.acquire()) {
+                        this.foldGeneration();
+                    }
+                    final boolean wasFoldingGeneration = foldingGeneration.getAndSet(false);
```

Comment:
I would feel better if this was in a finally clause (although if we have any exception here, we will fail the translog anyway). Thinking about this more - shall we move this logic into it's own "foldIfNeeded" method?